Documentation/designs/status-subresource.md

```diff
@@ -0,0 +1,436 @@
+kind: ServiceMonitor
+metadata:
+  name: example-app
+spec:
+  selector:
+    matchLabels:
+      app: example-app
+  endpoints:
+  - port: web
+status:
+  conditions:
+  - type: Valid
+    status: "True"
+    lastTransitionTime: "2022-02-08T23:54:22Z"
+  bindings:
+  - name: default/prometheus-agent-monitoring
+    prometheusRef:
+      resource: prometheusagents
+      name: monitoring
+      namespace: default
+    conditions:
+    - type: Reconciled
+      status: "True"
+      lastTransitionTime: "2022-02-08T23:54:22Z"
+```
+
+Example for an invalid ServiceMonitor (the secret reference doesn't exist):
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: example-app
+spec:
+  selector:
+    matchLabels:
+      app: example-app
+  endpoints:
+  - port: web
+    bearerTokenSecret:
+      name: metric-credentials
+      key: bearer-token
+status:
+  conditions:
+  - type: Valid
+    status: "False"
+    lastTransitionTime: "2022-02-08T23:54:22Z"
+    reason: InvalidSecretReference
+    message: "failed to get bearer token: failed to get token from secret: key 'bearer-token' in secret 'metric-credentials' not found"
+```
+
+Example for a ServiceMonitor that can't be bound to a Prometheus object:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1
+kind: ServiceMonitor
+metadata:
+  name: example-app
+spec:
+  selector:
+    matchLabels:
+      app: example-app
+  endpoints:
+  - port: web
+    bearerTokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
+status:
+  conditions:
+  - type: Valid
+    status: "True"
+    lastTransitionTime: "2022-02-08T23:54:22Z"
+  bindings:
+  - name: default/prometheus-agent-monitoring
+    prometheusRef:
+      resource: prometheusagents
+      name: monitoring
+      namespace: default
+    conditions:
+    - type: Reconciled
+      status: "False"
+      lastTransitionTime: "2022-02-08T23:54:22Z"
+      reason: DirectFileSystemAccessForbidden
+      message: "it accesses file system via bearer token file which Prometheus specification prohibits"
+```
+
+A given ServiceMonitor object may be selected by more than one
+Prometheus/PrometheusAgent resource which explains the presence of the
+`bindings` field. As explained in the [Kubernetes API
+conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#object-references), this creates additional complexity that need to be clarified:
+* The operator needs update permissions on the ServiceMonitor's status subresource.
+* When the referenced Prometheus resource stops selecting a given Service Monitor (`spec.serviceMonitorSelector` has changed or the ServiceMonitor's labels have been updated), the ServiceMonitor's status subresource should be updated.
+* When the referenced Prometheus resource is deleted, the ServiceMonitor's status subresource should be updated (this probably implies to setup a finalizer on the Prometheus resources).
+
+### Pod Monitor
+
+TBC
```

Comment:
I think it's fine to leave these sections out of the design and first implement the most used resources. I assume learnings from the implementation will influence the structure of the remaining status subresources.

Reply:
:+1: there's already a lot to explore/refine by just looking at the Prometheus and ServiceMonitor CRDs.